File: dotnet/src/Manager.Api/Features/Users/GetUserByEmail.cs
using Manager.Api.Common;
using Manager.Api.Common.Messaging;
using Manager.Api.Database;
using Microsoft.AspNetCore.Mvc;
using Microsoft.EntityFrameworkCore;
using Microsoft.Extensions.Caching.Hybrid;

namespace Manager.Api.Features.Users;

public static class GetUserByEmail
{
    public sealed record Query(string Email) : IQuery<Response>;

    public sealed record Response(long Id, string Name, string Email);

    internal sealed class Handler(
        ApplicationDbContext db,
        HybridCache cache) : IQueryHandler<Query, Response>
    {
        public async Task<Result<Response>> Handle(Query query, CancellationToken cancellationToken)
        {
            var response = await cache.GetOrCreateAsync(
                UserCacheKeys.ByEmail(query.Email),
                async ct =>
                {
                    return await db.Users
                        .AsNoTracking()
                        .Where(u => u.Email.ToLower() == query.Email.ToLower())
                        .Select(u => new Response(u.Id, u.Name, u.Email))
                        .SingleOrDefaultAsync(ct);
                },
                cancellationToken: cancellationToken);

            if (response is null)
            {
                await cache.RemoveAsync(UserCacheKeys.ByEmail(query.Email), cancellationToken);
                return Result.Failure<Response>(UserErrors.NotFound());
            }

            return Result.Success(response);
        }
    }

    public sealed class Endpoint : IEndpoint
    {
        public void MapEndpoint(IEndpointRouteBuilder app) =>
            app.MapGet(
                    "/api/v1/users/by-email",
                    async (
                        [FromQuery] string email,
                        [FromServices] IQueryHandler<Query, Response> handler,
                        CancellationToken cancellationToken) =>
                    {
                        var result = await handler.Handle(new Query(email), cancellationToken);
                        return result.Match(Results.Ok, CustomResults.Problem);
                    })
                .RequireAuthorization()
                .WithTags(Tags.Users);
    }
}

File: python/src/features/users/get_user_by_email.py
"""GetUserByEmail slice — fetch a user by exact email when authenticated."""

from __future__ import annotations

from dataclasses import dataclass

from fastapi import APIRouter, Depends, Query
from pydantic import BaseModel, ConfigDict
from sqlalchemy import func, select
from sqlalchemy.ext.asyncio import AsyncSession

from authentication.dependencies import get_current_subject
from common.cache import AppCache
from common.problem import problem_response
from common.result import Result
from database.models import UserModel
from database.session import get_db
from features.users.cache_keys import UserCacheKeys
from features.users.create_user import get_app_cache
from features.users.errors import UserErrors

router = APIRouter(prefix="/api/v1/users", tags=["Users"])

_CACHE_TTL_SECONDS = 60.0


@dataclass(frozen=True, slots=True)
class Query:
    email: str


@dataclass(frozen=True, slots=True)
class Response:
    id: int
    name: str
    email: str


class _UserNotFound(Exception):
    pass


class GetUserByEmailResponse(BaseModel):
    model_config = ConfigDict(populate_by_name=True)

    id: int
    name: str
    email: str


class Handler:
    def __init__(self, session: AsyncSession, cache: AppCache) -> None:
        self._session = session
        self._cache = cache

    async def handle(self, query: Query) -> Result[Response]:
        cache_key = UserCacheKeys.by_email(query.email)

        async def factory() -> Response:
            model = (
                await self._session.execute(
                    select(UserModel).where(
                        func.lower(UserModel.email) == query.email.lower()
                    )
                )
            ).scalar_one_or_none()
            if model is None:
                raise _UserNotFound()
            return Response(id=model.id, name=model.name, email=model.email)

        try:
            response = await self._cache.get_or_set(
                cache_key,
                factory,
                ttl_seconds=_CACHE_TTL_SECONDS,
            )
        except _UserNotFound:
            return Result.failure(UserErrors.not_found())

        return Result.success(response)


def _get_handler(
    session: AsyncSession = Depends(get_db),
    cache: AppCache = Depends(get_app_cache),
) -> Handler:
    return Handler(session, cache)


@router.get("/by-email", response_model=None)
async def get_user_by_email(
    email: str = Query(...),
    _: int = Depends(get_current_subject),
    handler: Handler = Depends(_get_handler),
) -> GetUserByEmailResponse | object:
    result = await handler.handle(Query(email=email))
    return result.match(
        on_success=lambda value: GetUserByEmailResponse(
            id=value.id,
            name=value.name,
            email=value.email,
        ),
        on_failure=problem_response,
    )
